talk/morelanguage/constexpr.tex

```diff
@@ -47,31 +45,67 @@
         return x > 1 ? x * f(x - 1) : 1;
       }
       static_assert(f(5)==120,"Expected f(5) to be 120!");
+      static_assert(f(5)==120); // C++17
     \end{cppcode*}
   \end{exampleblock}
 \end{frame}
 
 \begin{frame}[fragile]
   \frametitlecpp[14]{Generalized Constant Expressions(2)}
-   \begin{alertblock}{Few limitations in C++14 (more in C++11)}
-    \begin{itemize}
-    \item function's body cannot contain try-catch, uninitialized or static variables - details on \href{https://en.cppreference.com/w/cpp/language/constexpr}{cppreference}
-    \item arguments should be constexpr or literals in order to benefit from compile time computation
-    \end{itemize}
-  \end{alertblock}
   \begin{block}{Notes}
     \begin{itemize}
-    \item classes can have constexpr member functions
-    \item objects can be constexpr
+    \item A \mintinline{cpp}{constexpr} function \emph{may} be executed at compile time.
+    \begin{itemize}
+      \item Arguments must be \mintinline{cpp}{constexpr} or literals in order to benefit from compile time computation
+      \item Function body must be visible to the compiler
+    \end{itemize}
+    \item A \mintinline{cpp}{constexpr} function can also be used normally
+    \item A \mintinline{cpp}{constexpr} variable must be evaluated at compile time
+    \item Classes can have \mintinline{cpp}{constexpr} member functions
+    \item Objects used in constant expressions must be of \emph{literal type}:
       \begin{itemize}
-      \item if the constructor of their class is
+      \item an integral, floating-point, enum, reference, pointer type
+      \item a union (of at least one) or array of literal types
+      \item a class type with a \mintinline{cpp}{constexpr} constructor and
+            the destructor is trivial (or \mintinline{cpp}{constexpr} since C++20)
       \end{itemize}
-    \item a constexpr function can also be used normally
-    \item but a constexpr variable has to be evaluated at compile time
+    \item A constexpr function is implicitly \mintinline{cpp}{inline} (header files)
     \end{itemize}
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile,shrink=15]
+  \frametitlecpp[11]{Limitations of constexpr functions}
+   \begin{alertblock}{C++11}
+     \begin{itemize}
+     \item Non-virtual function with basically a single return statement
```

Comment:
```suggestion
     \item Non-virtual function with a single return statement
```

Reply:
I put the `basically` because you could also have e.g. an alias declaration before the return statement. But yeah, let's just make it simple on the slides.